File: 2022/python/day03.py
import os

INPUT_TEST = f"""vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw
"""
letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"


with open("../input/day03.txt", "r") as handle:
    content = handle.readlines()


def part1():
    idx_sum = 0

    for line in content:
        if len(line) > 1:
            idx_half = int(len(line) / 2)
            first = line[:idx_half]
            second = line[idx_half:]
            dup = [c for c in letters if (c in first and c in second)]
            idx_sum += letters.index(dup[0]) + 1

    print(idx_sum)


def part2():
    num_groups = int((len(content) - 1) / 3)
    group_priorities = 0

    for i in range(num_groups):
        common = [
            c
            for c in letters
            if (
                c in content[i * 3]
                and c in content[i * 3 + 1]
                and c in content[i * 3 + 2]
            )
        ]
        group_priorities += letters.index(common[0]) + 1

    print(group_priorities)


if __name__ == "__main__":
    part1()
    part2()

File: 2022/rust/src/day03.rs
use std::{fs, println};

const LETTERS: &'static str = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

fn main() {
    let input = fs::read_to_string("../input/day03.txt").unwrap();

    let mut priority_sum = 0;
    let mut priority_group = 0;

    // part 1
    for line in input.lines() {
        let num_char = line.len() / 2;
        let first = line[0..num_char].to_string();
        let second = line[num_char..].to_string();

        for char in LETTERS.chars() {
            if first.contains(char) && second.contains(char) {
                priority_sum += 1 + LETTERS.chars().position(|c| c == char).unwrap();
            }
        }
    }

    // part 2
    let lines: Vec<&str> = input.lines().collect();
    let num_groups = lines.len() / 3;
    for i in 0..num_groups {
        for char in LETTERS.chars() {
            if lines[i * 3].contains(char)
                && lines[i * 3 + 1].contains(char)
                && lines[i * 3 + 2].contains(char)
            {
                priority_group += 1 + LETTERS.chars().position(|c| c == char).unwrap();
            }
        }
    }
    println!("day03-part1: {}", priority_sum);
    println!("day03-part2: {}", priority_group);
}
